### DISCUSSION TITLE
Are we using -bpf-expand-memcpy-in-order yet?

### DISCUSSION THREAD

**Discussion by Birch-san** [bpftrace/bpftrace#1800] [Ideas]
On big strings https://github.com/iovisor/bpftrace/pull/1360, I [encountered](https://github.com/iovisor/bpftrace/pull/1360#issuecomment-646502771):

```
A call to built-in function 'memset' is not supported.
```

this specifically would happen if you emitted more than 128 store instructions… for example if you tried to do a 64-bit-aligned memset/memcpy on a string of length 1024, or an 8-bit aligned memset/memcpy on a string of length 128.

@mmarchini encountered the same (I believe) on https://github.com/iovisor/bpftrace/pull/750:

> memset will only work for small values (even if we're using a map). I'm on another branch right now, I'll try to share more information later and I'll add a comment here explaining why this is implemented this way.

we'd figured "I guess we'll have to use probe_read() instead, if we want to do a memset/memcpy of that size". that had downsides — you have to verify whether it succeeded, which means a conditional branch and a plan of how to abort.

but I think I see a better option now, `-bpf-expand-memcpy-in-order`:

https://reviews.llvm.org/rG71d81e5c8fcebb7663973d686ac1ad3309502f4b

our IR would remain as the LLVM intrinsic memset/memcpy, and only be expanded at a later optimization pass.

if this flag removes the limit: it would be **very** good news for big strings (emitting probe_reads was the change that impacted the IR most)

**Comment by Birch-san** [comment]
@yonghong-song sorry for direct approach, but do you know whether it's possible for bpftrace to programmatically enable `bpf-expand-memcpy-in-order`?

ordinarily, it's enabled via the command-line, i.e. `llc -bpf-expand-memcpy-in-order`. but since bpftrace compiles/links against the LLVM library, we cannot pass command-line options like this.

I tried passing command-line options using `llvm::cl::ParseCommandLineOptions`:  
https://github.com/iovisor/bpftrace/compare/master...Birch-san:expand-memcpy-in-order?expand=1

but my test program still explodes when I attempt to emit more than 128 8-byte store instructions (i.e. `MaxStoresPerMemset`):

```
sudo BPFTRACE_STRLEN=64 ./src/bpftrace -e 'BEGIN { $a = str($1) }' 'hey'
Attaching 1 probe...

# memset() of 1024 bytes is allowed (we exceed our stack, but this is a known and unrelated problem)
sudo BPFTRACE_STRLEN=1024 ./src/bpftrace -e 'BEGIN { $a = str($1) }' 'hey'
error: <unknown>:0:0: in function BEGIN i64 (i8*): Looks like the BPF stack limit of 512 bytes is exceeded. Please move large on stack variables into BPF per-cpu array map.

# memset() of 1032 bytes is disallowed by LLVM BPF backend
sudo BPFTRACE_STRLEN=1032 ./src/bpftrace -e 'BEGIN { $a = str($1) }' 'hey'
error: <unknown>:0:0: in function BEGIN i64 (i8*): A call to built-in function 'memset' is not supported.
```

so probably my trick to pass command-line options doesn't work. is there anything else we can do to enable `bpf-expand-memcpy-in-order`?

and am I understanding correctly that this option can fix "A call to built-in function 'memset' is not supported" for large memsets?

**Comment by danobi** [comment]
My guess is `-bpf-expand-memcpy-in-order` only applies to `memcpy()`, not `memset()` despite the reference in the commit. How about as an alternative, we implement `b_.CREATE_MEMSET()` as a bounded loop instead of calling an intrinsic (`CreateMemSet()`)? As long as we provide a constant bound on the loop the verifier should be happy.

**Reply by danobi** [reply]
If the memset()  length is short enough we can keep calling the intrinsic too.

**Reply by Birch-san** [reply]
hm.. certainly that commit refers almost exclusively to memcpy()…

when I was playing around with big strings, one thing I tried was a C++ loop to emit multiple small `memset()`/`memcpy()`s. but I got the same result as though I'd emitted one big `memset()`/`memcpy()` ("A call to built-in function 'memset' is not supported."). this may be the "expand" to which it refers? `memset(128 x u64), memset(128 x u64)` expands to 256 u64 stores, which exceeds `MaxStoresPerMemset`?

I didn't try a _BPF_ bounded loop. I think that would _work_, but would the LLVM IR be messy? does the creation of a BPF loop entail the creation of additional LLVM blocks, or invocation of BPF helpers?

if BPF loops involve BPF helpers, presumably we'd have to use a condition to check success — if so, this would put us in the same place as `probe_read()`. but if _not_, then that's an improvement. either way: at least avoids having to mock out a memory address in test fixtures.

**Reply by danobi** [reply]
> when I was playing around with big strings, one thing I tried was a C++ loop to emit multiple small memset()/memcpy()s. but I got the same result as though I'd emitted one big memset()/memcpy() ("A call to built-in function 'memset' is not supported."). this may be the "expand" to which it refers? memset(128 x u64), memset(128 x u64) expands to 256 u64 stores, which exceeds MaxStoresPerMemset?

Maybe LLVM is being smart and decided to fuse the memsets. Just a guess.

> I didn't try a BPF bounded loop. I think that would work, but would the LLVM IR be messy? does the creation of a BPF loop entail the creation of additional LLVM blocks, or invocation of BPF helpers?

There'll be some more blocks, yeah. It also might not work b/c LLVM is too clever for us. But I think it's worth a try. Maybe LLVM will only unroll the loop but not realize it's actually a memset.

There's no helpers involved, just a plain old loop.

Long term I'm hoping https://github.com/iovisor/bpftrace/issues/1845 helps us avoid LLVM trickery.

**Reply by Birch-san** [reply]
> decided to fuse the memsets

yup, I reckon that's what we were seeing. which would mean a C++ loop emitting multiple memsets couldn't work.

> There'll be some more blocks, yeah.

hmm that's a shame. I dislike solutions where "big memset" produces drastically different IR to "small memset", because then you have to decide "how are we going to test this" — do you make duplicates of affected test fixtures (which is most of them), for `BPFTRACE_STRLEN=(more than)1024`?

> worth a try

certainly I'd be interested to _know_ whether it works. yeah, maybe LLVM would unroll the loop, and maybe it would do so without fusing memsets. but in terms of complexity of IR: this has most of the reviewability/testability problems of the `probe_read` solution. well, maybe slightly better insofar as you don't need to "validate whether helper succeeded; exit program on failure", and that there's no addresses that need mocking in tests.

the other question we could be asking is "can we tell LLVM not to fuse memsets" — in which case the simpler C++ loop becomes an option again. Maybe we could add our own pass to split memsets!

but regardless, I think there's a _perf_ reason why we'd want `-bpf-expand-memcpy-in-order`. I was finding that if I set `BPFTRACE_STRLEN` high (e.g. to the limit, 1024): it made tools (like tcpconnect.bt) take a super long time (tens of seconds) to attach probes. I don't know where the time was spent (C++, LLVM compiler, kernel BPF verifier, kernel BPF JIT) but it was clear that longer strings meant longer startup time. I think specifically it's the number of load/stores, since the slowdown was alleviated _substantially_ by changing memset/memcpy to be 8-byte-aligned. perhaps this is the problem referenced in the [`-bpf-expand-memcpy-in-order` changeset](https://reviews.llvm.org/rG71d81e5c8fcebb7663973d686ac1ad3309502f4b)? that expanding memcpy into load/store pairs early, means that hundreds more instructions will be in-scope for the optimization passes that follow. and if those optimization passes _move_ the load/store pairs: they may no longer fit a pattern that the kernel BPF JIT can optimize.

regarding https://github.com/iovisor/bpftrace/issues/1845… are you saying we could approach "big strings" as being "a feature supported only by the new backend"?

**Reply by danobi** [reply]
> do you make duplicates of affected test fixtures (which is most of them), for BPFTRACE_STRLEN=(more than)1024?

What if instead of using llvm's builtin memset we codegen memset to be a:
```
def memset(addr, val, n):
  for i in range(n):
    addr[i] = val
```
That would unify all the fixtures, right?

> Maybe we could add our own pass to split memsets!

Hmm, that sounds pretty complex given how fast LLVM internals move. If it's pretty simple that would be great but I suspect it'd just be a massive PITA to maintain across multiple LLVM versions.

> but regardless, I think there's a perf reason why we'd want -bpf-expand-memcpy-in-order. I was finding that if I set BPFTRACE_STRLEN high (e.g. to the limit, 1024): it made tools (like tcpconnect.bt) take a super long time (tens of seconds) to attach probes. I don't know where the time was spent (C++, LLVM compiler, kernel BPF verifier, kernel BPF JIT) but it was clear that longer strings meant longer startup time.

I suspect emiting our own memset wouldn't be too much work to try out (idk, 1-2 hours?). If it's noticeably slower for even small strings then yeah, maybe our own memset is a no-go for perf reasons.

> regarding https://github.com/iovisor/bpftrace/issues/1845… are you saying we could approach "big strings" as being "a feature supported only by the new backend"?

We _could_, but that's not what I'm suggesting. This feels to me like we're letting the perfect be the enemy of the good. Big strings are definitely useful and it seems like at some point we'll have to compromise and eat some complexity. We could do it the "best" way in LLVM upstream but that takes a while to percolate to all users. Plus there's the question of how we detect llvm even has that "feature"/bug fix.

IMO some churn in codegen fixtures is OK. Whatever gets us to big strings with the least complexity and easiest maintainability I would be in favor of.

**Reply by Birch-san** [reply]
> instead of using llvm's builtin memset

okay yeah, I think that could work. yes, probably pretty easy to do. ideally would emit 8-byte load/stores where possible, to emit fewer instructions. but since we currently do 1-byte aligned memset/memcpy everywhere: maybe further preparations are needed for that optimization.

I'm optimistic the perf would be comparable for small strings (since it sounds like LLVM expands memset/memcpy into load/store instructions pretty early). and I anticipate that long strings would experience a similar slow startup to what I see in the current code (because you end up in the same situation — thousands of load/stores being subjected to several optimization passes).

but overall, seems like a solution that's easy to implement/maintain, and easy to review/test (i.e. small strings work the same as big strings, so you don't need separate testing).

admittedly every memset/memcpy in test fixtures would get expanded into potentially many load/store instructions — you lose brevity and semantic intention — but there are upsides to this, in that the IR would look a bit more like the real BPF program.

**Reply by danobi** [reply]
> admittedly every memset/memcpy in test fixtures would get expanded into potentially many load/store instructions

Not if you emit a bounded loop, right? Emitting thousands of stores would be pretty bad b/c there's an instruction limit for progs. But I think a bounded loop should be ok.

**Reply by Birch-san** [reply]
> Not if you emit a bounded loop, right?

ah, I'd thought by "write our own memset" we were discussing using a C++ loop to emit lots of LLVM store/loads. hadn't considered instruction limit. 1mill instructions admittedly equips us to solve kilobyte problems, but puts megabytes out-of-reach.

then, maybe that steers us to BPF bounded loop as the most futureproof solution. since I think even if we removed the "invalid memset" limit on LLVM: it'd still be emitting thousands of load/stores. bounded loop may even solve the startup speed problem for large strings, if that's related at all to "how many load/stores exist in the LLVM IR".

would we resort to BPF bounded loop even for small strings? would that emit less efficient BPF bytecode (i.e. with condition, accumulator and jumps)? mainly wondering because if big strings produce different bytecode, then we'd need a plan of how to test it — in the worst case this means "big string" variants of every existing test.

would also be good to know "what kinds of optimizations can the kernel BPF JIT perform" (could it unroll our loops?)

**Reply by danobi** [reply]
> would we resort to BPF bounded loop even for small strings?

That's one option. Another option is to use builtin memset for N <= 1024-ish and use a bounded loop for N > 1024-ish (which I think llvm is doing anyways, which is why it's disabled in BPF backend). Now that I typed that out, I kinda like that idea b/c then we don't churn the fixtures.

> would that emit less efficient BPF bytecode (i.e. with condition, accumulator and jumps)?

Yeah, probably, but we can be moderately clever too and do 8 byte strides. It's not like BPF instruction set is that complicated so I doubt there's that much room for extremely complicated memset optimizations.

> then we'd need a plan of how to test it — in the worst case this means "big string" variants of every existing test.

I don't think we need to resort to "big string" variants of every test. Our codegen is pretty modular so if we have confidence that our memset implementation is working well in isolation I'm pretty confident it'll play nice with the other parts of codegen.

memset is also a very well-trodden code path for bpftrace progs so if there's any issues we'll probably discover it really soon.

**Reply by Birch-san** [reply]
okay cool; testing it in isolation makes it substantially easier.

8-byte strides would be good. but it makes me wonder why we're currently doing 1-byte-aligned memsets. if there's an underlying reason, then it may impact our ability to do 8-byte strides.  
in my "big strings" branch: I tried changing every memset/memcpy to be 8-byte-aligned and it worked for the minimal programs I tried it on. well, maybe it failed on a tools test? I think memset supports "use the alignment of the pointed-to datatype" and memcpy supports "look at both pointed-to datatypes, and pick whichever alignment is smaller" — which we could make use of in more places if we used structs rather than Int8Arrays in more places.
